src/components/MultiHopTrade/components/SharedTradeInput/SharedRecipientAddress.tsx

```diff
@@ -0,0 +1,387 @@
+  buyAsset,
+  manualAddressEntryDescription,
+}: ManualRecipientAddressLabelsProps) => {
+  const translate = useTranslate()
+  const wallet = useWallet().state.wallet
+  const { isSnapInstalled } = useIsSnapInstalled()
+  const { open: openSnapsModal } = useModal('snaps')
+  const { open: openManageAccountsModal } = useModal('manageAccounts')
+
+  const walletSupportsBuyAssetChainAtRuntime = useWalletSupportsChainAtRuntime(
+    buyAsset.chainId,
+    wallet,
+  )
+
+  const buyAssetChainName = useMemo(() => {
+    const chainAdapterManager = getChainAdapterManager()
+    return chainAdapterManager.get(buyAsset.chainId)?.getDisplayName()
+  }, [buyAsset])
+
+  const isMetaMaskWalletWithoutSnap = useMemo(() => {
+    return !isSnapInstalled && wallet && isMetaMask(wallet)
+  }, [isSnapInstalled, wallet])
+
+  // We're enabling the snap, so no versioning concerns here
+  const handleEnableShapeShiftSnap = useCallback(() => openSnapsModal({}), [openSnapsModal])
+  const handleAddAccount = useCallback(() => openManageAccountsModal({}), [openManageAccountsModal])
+
+  return (
+    <>
+      <FormLabel color='yellow.400'>
+        {manualAddressEntryDescription ??
+          translate('trade.receiveAddressDescription', { chainName: buyAssetChainName })}
+        {isMetaMaskWalletWithoutSnap && (
+          <Link textDecor='underline' ml={1} onClick={handleEnableShapeShiftSnap}>
+            {translate('trade.or')}
+            &nbsp;{translate('trade.enableMetaMaskSnap')}
+          </Link>
+        )}
+        {walletSupportsBuyAssetChainAtRuntime && (
+          <>
+            &nbsp;
+            {translate('common.or')}
+            <Link textDecor='underline' ml={1} onClick={handleAddAccount}>
+              {translate('trade.connectChain', { chainName: buyAssetChainName })}
+            </Link>
+          </>
+        )}
+        &nbsp;{translate('trade.toContinue')}
+      </FormLabel>
+      <FormLabel color='white.500' w='full' fontWeight='bold'>
+        {translate('trade.receiveAddress')}
+      </FormLabel>
+    </>
+  )
+}
+
+type SharedRecipientAddressProps = {
+  buyAsset: Asset
+  customRecipientAddressDescription?: string
+  isWalletReceiveAddressLoading: boolean
+  manualAddressEntryDescription?: string
+  manualReceiveAddress: string | undefined
+  shouldForceManualAddressEntry?: boolean
+  walletReceiveAddress: string | undefined
+  onCancel: () => void
+  onEdit: () => void
+  onError: () => void
+  onIsValidatingChange: (isValidating: boolean) => void
+  onIsValidChange: (isValid: boolean) => void
+  onReset: () => void
+  onSubmit: (address: string) => void
+}
+
+export const SharedRecipientAddress = ({
+  buyAsset,
+  customRecipientAddressDescription,
+  isWalletReceiveAddressLoading,
+  manualAddressEntryDescription,
+  manualReceiveAddress,
+  shouldForceManualAddressEntry,
+  walletReceiveAddress,
+  onCancel,
+  onEdit,
+  onError,
+  onIsValidatingChange,
+  onIsValidChange,
+  onReset,
+  onSubmit,
```

Comment:
nitpick: I think these should all be `handle` prefixed, too

Reply:
Apparently props need to use `onXyz` and variables to use `handleXyz`
https://react.dev/learn/responding-to-events

So a new handler is `const handleXyz = ...` and it's passed in like `onXyz={handleXyz}`

If a prop is drilled directly into another component, it's ok to do `onClick={onClick}`

Reply:
Excellent 🙏